File: example_py/seer_basic_example.py
#!/usr/bin/env python3

import argparse
import logging
import signal
import time

import flexiv_amr


logger = logging.getLogger("seer_basic_example")
stop_requested = False


def setup_logging() -> None:
    logging.basicConfig(
        level=logging.INFO,
        format="[%(asctime)s.%(msecs)03d] [%(levelname)s] %(message)s",
        datefmt="%Y-%m-%d %H:%M:%S",
    )


def signal_handler(signum, frame) -> None:
    global stop_requested
    stop_requested = True
    logger.info("Stop signal received.")


def create_seer_options(amr_ip: str) -> flexiv_amr.SeerAmrOptions:
    options = flexiv_amr.SeerAmrOptions()

    options.host = amr_ip
    options.auto_connect = False
    options.auto_reconnect = True

    # Optional: adjust timeout settings if needed.
    # options.time_out.connect_timeout_ms = 3000
    # options.time_out.send_timeout_ms = 3000
    # options.time_out.recv_timeout_ms = 3000

    return options


def print_amr_info(amr: flexiv_amr.SeerAmrClient) -> None:
    info = amr.get_amr_info()

    logger.info("=== AMR Info ===")
    logger.info("AMR ID: %s", info.amr_id)
    logger.info("AMR Name: %s", info.amr_name)
    logger.info("Model: %s", info.amr_model_name)
    logger.info("Version: %s", info.robokit_version)
    logger.info("Current Map: %s", info.current_map)


def print_runtime_status(amr: flexiv_amr.SeerAmrClient, loop_count: int) -> None:
    logger.info("========== Loop %d ==========", loop_count)

    pose = amr.get_amr_pose()
    logger.info(
        "Pose: x=%.3f, y=%.3f, angle=%.3f, confidence=%.3f, "
        "localization_method=%s, current_station=%s, last_station=%s",
        pose.x,
        pose.y,
        pose.angle,
        pose.confidence,
        pose.localization_method,
        pose.current_station,
        pose.last_station,
    )

    battery = amr.get_battery_status()
    logger.info(
        "Battery: level=%.1f, voltage=%.3f, current=%.3f, charging=%s",
        battery.battery_level,
        battery.voltage,
        battery.current,
        battery.charging,
    )

    navigation = amr.get_navigation_status()
    logger.info(
        "Navigation: task_status=%s, task_type=%s, target_id=%s",
        navigation.task_status,
        navigation.task_type,
        navigation.target_id,
    )

    task = amr.get_task_status()
    logger.info(
        "Task: percentage=%.1f, distance=%.3f, closest_target=%s, "
        "source_name=%s, target_name=%s, task_states.size=%d",
        task.percentage,
        task.distance,
        task.closest_target,
        task.source_name,
        task.target_name,
        len(task.task_states),
    )

    location_status = amr.get_location_status()
    logger.info("Location: reloc_status=%s", location_status)

    ultrasonic = amr.get_ultrasonic_data()
    logger.info("Ultrasonic: %d nodes", len(ultrasonic.ultrasonic_nodes))
    for node in ultrasonic.ultrasonic_nodes:
        logger.info(
            "Ultrasonic node: id=%s, dist=%.3f m, valid=%s",
            node.id,
            node.dist,
            node.valid,
        )

    encoder_data = amr.get_encoder_data()
    for motor_encoder in encoder_data.motor_encoders:
        logger.info("Motor encoder: %s", motor_encoder)
    if encoder_data.legacy_encoder_pulses:
        logger.info("Legacy encoder pulses: %s", encoder_data.legacy_encoder_pulses)

    imu = amr.get_imu_data()
    logger.info(
        "IMU: yaw=%.6f, roll=%.6f, pitch=%.6f, "
        "acc=[%.3f, %.3f, %.3f], rot=[%.3f, %.3f, %.3f]",
        imu.yaw,
        imu.roll,
        imu.pitch,
        imu.acc_x,
        imu.acc_y,
        imu.acc_z,
        imu.rot_x,
        imu.rot_y,
        imu.rot_z,
    )

    logger.info("IsConnected: %s", amr.is_connected())


def main() -> int:
    setup_logging()

    parser = argparse.ArgumentParser(description="Basic Python example for SEER AMR.")
    parser.add_argument(
        "amr_ip",
        help="SEER AMR IP address, for example: 192.168.192.5",
    )
    parser.add_argument(
        "interval_ms",
        nargs="?",
        type=int,
        default=1000,
        help="Polling interval in milliseconds. Default: 1000",
    )

    args = parser.parse_args()

    signal.signal(signal.SIGINT, signal_handler)
    signal.signal(signal.SIGTERM, signal_handler)

    options = create_seer_options(args.amr_ip)

    logger.info("connect_timeout_ms: %d", options.time_out.connect_timeout_ms)
    logger.info("send_timeout_ms: %d", options.time_out.send_timeout_ms)
    logger.info("recv_timeout_ms: %d", options.time_out.recv_timeout_ms)

    amr = flexiv_amr.SeerAmrClient(options)

    try:
        logger.info("Connecting to SEER AMR: %s ...", args.amr_ip)
        amr.connect()

        logger.info("Connected: %s", amr.is_connected())

        print_amr_info(amr)

        loop_count = 0

        while not stop_requested:
            loop_count += 1

            try:
                print_runtime_status(amr, loop_count)

            except flexiv_amr.AmrException as e:
                logger.error("AMR exception in loop %d: %s", loop_count, e)
                logger.warning("Breaking loop because AMR communication failed.")
                break

            except Exception as e:
                logger.exception("Unexpected exception in loop %d: %s", loop_count, e)
                logger.warning("Breaking loop because unexpected exception occurred.")
                break

            time.sleep(args.interval_ms / 1000.0)

        logger.info("Disconnecting...")
        amr.disconnect()
        logger.info("Disconnected.")

        return 0

    except flexiv_amr.AmrException as e:
        logger.error("AMR exception: %s", e)

    except Exception as e:
        logger.exception("Unexpected exception: %s", e)

    finally:
        try:
            if amr.is_connected():
                amr.disconnect()
        except Exception:
            pass

    return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: example/seer_basic_example.cpp
/**
 * @example seer_basic_example.cpp
 * This example connects to a SEER AMR and prints basic status information.
 */

#include <atomic>
#include <chrono>
#include <csignal>
#include <iostream>
#include <string>
#include <thread>

#include <flexiv/amr/exceptions.h>
#include <flexiv/amr/vendor/seer/seer_amr.h>
#include <spdlog/spdlog.h>

namespace {
volatile std::sig_atomic_t g_stop_requested = 0;

void SignalHandler(int) {
    g_stop_requested = 1;
}
}  // namespace

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage:\n"
                  << "  " << argv[0] << " <amr_ip> [interval_ms]\n\n"
                  << "Example:\n"
                  << "  " << argv[0] << " 192.168.192.5 1000\n";
        return 1;
    }

    std::signal(SIGINT, SignalHandler);
    std::signal(SIGTERM, SignalHandler);

    const std::string amr_ip = argv[1];
    const int interval_ms = argc >= 3 ? std::stoi(argv[2]) : 1000;

    spdlog::set_pattern("[%Y-%m-%d %H:%M:%S.%e] [%^%l%$] %v");

    try {
        flexiv::amr::seer::SeerAmrOptions options;
        options.host = amr_ip;
        options.auto_connect = false;

        spdlog::info("connect_timeout_ms: {}", options.time_out.connect_timeout_ms);
        spdlog::info("send_timeout_ms: {}", options.time_out.send_timeout_ms);
        spdlog::info("recv_timeout_ms: {}", options.time_out.recv_timeout_ms);

        flexiv::amr::seer::SeerAmrClient amr(options);

        spdlog::info("Connecting to SEER AMR: {} ...", amr_ip);
        amr.Connect();

        spdlog::info("Connected: {}", amr.IsConnected());

        {
            const auto info = amr.GetAmrInfo();

            spdlog::info("=== AMR Info ===");
            spdlog::info("AMR ID: {}", info.amr_id);
            spdlog::info("AMR Name: {}", info.amr_name);
            spdlog::info("Model: {}", info.amr_model_name);
            spdlog::info("Version: {}", info.robokit_version);
            spdlog::info("Current Map: {}", info.current_map);
        }

        std::size_t loop_count = 0;

        while (!g_stop_requested) {
            ++loop_count;

            try {
                spdlog::info("========== Loop {} ==========", loop_count);

                const auto pose = amr.GetAmrPose();
                spdlog::info("Pose: x={:.3f}, y={:.3f}, angle={:.3f}, confidence={:.3f}, "
                             "localization_method={}, current_station={}, last_station={}",
                             pose.x,
                             pose.y,
                             pose.angle,
                             pose.confidence,
                             pose.localization_method,
                             pose.current_station,
                             pose.last_station);

                const auto battery = amr.GetBatteryStatus();
                spdlog::info("Battery: level={:.1f}, voltage={:.3f}, current={:.3f}, charging={}",
                             battery.battery_level,
                             battery.voltage,
                             battery.current,
                             battery.charging);

                const auto navigation = amr.GetNavigationStatus();
                spdlog::info("Navigation: task_status={}, task_type={}, target_id={}",
                             navigation.task_status,
                             navigation.task_type,
                             navigation.target_id);

                const auto task = amr.GetTaskStatus();
                spdlog::info("Task: percentage={:.1f}, distance={:.3f}, closest_target={}, "
                             "source_name={}, target_name={}, task_states.size={}",
                             task.percentage,
                             task.distance,
                             task.closest_target,
                             task.source_name,
                             task.target_name,
                             task.task_states.size());

                const int location_status = amr.GetLocationStatus();
                spdlog::info("Location: reloc_status={}", location_status);

                const auto ultrasonic = amr.GetUltrasonicData();
                spdlog::info("Ultrasonic: {} nodes", ultrasonic.ultrasonic_nodes.size());
                for (const auto& node : ultrasonic.ultrasonic_nodes) {
                    spdlog::info("Ultrasonic node: id={}, dist={:.3f} m, valid={}",
                                 node.id,
                                 node.dist,
                                 node.valid);
                }

                const auto encoder_data = amr.GetEncoderData();
                for (const auto& motor_encoder : encoder_data.motor_encoders) {
                    spdlog::info("Motor encoder: {}", motor_encoder.dump());
                }
                if (!encoder_data.legacy_encoder_pulses.empty()) {
                    spdlog::info("Legacy encoder pulses: {} values",
                                 encoder_data.legacy_encoder_pulses.size());
                }

                const auto imu = amr.GetImuData();
                spdlog::info("IMU: yaw={:.6f}, roll={:.6f}, pitch={:.6f}, "
                             "acc=[{:.3f}, {:.3f}, {:.3f}], rot=[{:.3f}, {:.3f}, {:.3f}]",
                             imu.yaw,
                             imu.roll,
                             imu.pitch,
                             imu.acc_x,
                             imu.acc_y,
                             imu.acc_z,
                             imu.rot_x,
                             imu.rot_y,
                             imu.rot_z);

                spdlog::info("IsConnected: {}", amr.IsConnected());
            } catch (const flexiv::amr::AmrException& e) {
                spdlog::error("AMR exception in loop {}: {}, code: {}",
                              loop_count,
                              e.what(),
                              static_cast<int>(e.code()));

                spdlog::warn("Breaking loop because AMR communication failed.");
                break;
            } catch (const std::exception& e) {
                spdlog::error("std exception in loop {}: {}", loop_count, e.what());
                spdlog::warn("Breaking loop because unexpected exception occurred.");
                break;
            }

            std::this_thread::sleep_for(std::chrono::milliseconds(interval_ms));
        }

        spdlog::info("Disconnecting...");
        amr.Disconnect();
        spdlog::info("Disconnected.");

        return 0;
    } catch (const flexiv::amr::AmrException& e) {
        spdlog::error("AMR exception: {}, code: {}", e.what(), static_cast<int>(e.code()));
    } catch (const std::exception& e) {
        spdlog::error("std exception: {}", e.what());
    }

    return 1;
}
